unblob/processing.py

```diff
@@ -136,76 +143,91 @@ def _process_task(self, result: TaskResult, task: Task):
-                if task.depth < self._entropy_depth:
-                    calculate_entropy(task.path, draw_plot=self._entropy_plot)
-                return
+                self._calculate_entropies([self.task.path])
 
-            extract_dir = make_extract_dir(task.root, task.path, self._extract_root)
+    def _process_chunks(
+        self, file, outer_chunks: List[ValidChunk], unknown_chunks: List[UnknownChunk]
+    ):
+        extract_dir = make_extract_dir(
+            self.task.root, self.task.path, self.parameters.extract_root
+        )
 
-            carved_unknown_paths = carve_unknown_chunks(
-                extract_dir, file, unknown_chunks
-            )
-            if task.depth < self._entropy_depth:
-                for carved_unknown_path in carved_unknown_paths:
-                    calculate_entropy(carved_unknown_path, draw_plot=self._entropy_plot)
-
-            for chunk in outer_chunks:
-                carved_valid_path = carve_valid_chunk(extract_dir, file, chunk)
-
-                if chunk.is_encrypted:
-                    logger.warning(
-                        "Do not attempt to extract encrypted file",
-                        path=carved_valid_path,
-                        chunk=chunk,
-                    )
-                    continue
+        carved_unknown_paths = carve_unknown_chunks(extract_dir, file, unknown_chunks)
+        self._calculate_entropies(carved_unknown_paths)
 
-                inpath, outdir = get_extract_paths(extract_dir, carved_valid_path)
+        for chunk in outer_chunks:
+            self._extract_chunk(extract_dir, file, chunk)
 
-                extractor = chunk.handler.EXTRACTOR
-                if extractor is not None:
-                    self._extract_chunk(task, inpath, outdir, extractor, result)
+    def _calculate_entropies(self, paths: List[Path]):
+        if self.task.depth < self.parameters.entropy_depth:
+            for path in paths:
+                calculate_entropy(path, draw_plot=self.parameters.entropy_plot)
 
     def _extract_chunk(
         self,
-        task: Task,
-        inpath: Path,
-        outdir: Path,
-        extractor: Extractor,
-        result: TaskResult,
+        extract_dir: Path,
+        file,
+        chunk: ValidChunk,
     ):
-        # We only extract every blob once, it's a mistake to extract the same blog again
-        outdir.mkdir(parents=True, exist_ok=False)
+        carved_path = carve_valid_chunk(extract_dir, file, chunk)
+        inpath, outdir = get_extract_paths(extract_dir, carved_path)
+
         try:
-            extractor.extract(inpath, outdir, result)
+            chunk.extract(inpath, outdir, self.result)
+            if not self.parameters.keep_extracted_chunks:
+                logger.debug("Removing extracted chunk", path=inpath)
+                inpath.unlink()
+
+        except ExtractError:
+            # already reported & logged by extractor
```

Comment:
Another indication that this needs a better name but I am out of ideas.

Another possible approach: have this error accept an argument that will be added to the TaskResult at this level. I guess it would make the expected and unexpected error handling more streamlined: 
- Got an ExtractError(reports)? Good, let's append it to the TaskResult
- Got any other (unexpected) error? Not a problem, let's convert it to an UnknownErrorReport

Reply:
implemented :)